cmd/package_managers.go

```diff
@@ -138,3 +205,145 @@ func fetchGitRepositoryFromRubyGems(packageName string, manager packageManagerCl
+		"PackageBaseAddress/3.0.0")
+	if err != nil {
+		return "", err
+	}
+	nugetRegistrationBaseURL, err := getFieldFromIndexResults(nugetIndexResults.Resources,
+		"RegistrationsBaseUrl/3.6.0")
+	if err != nil {
+		return "", err
+	}
+
+	lowerCasePackageName := strings.ToLower(packageName)
+	lastPackageVersion, err := getLatestListedVersion(nugetRegistrationBaseURL,
+		lowerCasePackageName, manager)
+	if err != nil {
+		return "", err
+	}
+
+	respPackageSpec, err := manager.Get(
+		nugetPackageBaseURL+"%[1]v/"+lastPackageVersion+"/%[1]v.nuspec", lowerCasePackageName)
+	if err != nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal,
+			fmt.Sprintf("failed to get nuget package spec json: %v", err))
+	}
+	defer respPackageSpec.Body.Close()
+
+	packageSpecResults := &nugetNuspec{}
+	err = xml.NewDecoder(respPackageSpec.Body).Decode(packageSpecResults)
+
+	if err != nil || packageSpecResults.Metadata == (nuspecMetadata{}) {
+		return "", sce.WithMessage(sce.ErrScorecardInternal,
+			fmt.Sprintf("failed to parse nuget nuspec xml: %v", err))
+	}
+	projectURL := ""
+	if packageSpecResults.Metadata.Repository != (nuspecRepository{}) &&
+		len(strings.TrimSpace(packageSpecResults.Metadata.Repository.URL)) != 0 {
+		projectURL = packageSpecResults.Metadata.Repository.URL
+	}
+	if len(strings.TrimSpace(packageSpecResults.Metadata.ProjectURL)) != 0 {
+		projectURL = packageSpecResults.Metadata.ProjectURL
+	}
+	if isSupportedProjectURL(projectURL) {
+		return strings.TrimSuffix(projectURL, ".git"), nil
+	}
+	return "", sce.WithMessage(sce.ErrScorecardInternal,
+		fmt.Sprintf("source repo is not defined for nuget package %v", packageName))
+}
+
+func isSupportedProjectURL(projectURL string) bool {
+	pattern := `^(?:https?://)?(?:www\.)?(?:github|gitlab)\.com/([A-Za-z0-9_\.-]+)/([A-Za-z0-9_\./-]+)$`
+	regex := regexp.MustCompile(pattern)
+	return regex.MatchString(projectURL)
+}
+
+// Gets the latest listed nuget version of a package, based on the protocol defined at
+// https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource#enumerate-package-versions
+func getLatestListedVersion(baseURL, packageName string, manager packageManagerClient) (string, error) {
+	resPackageRegistrationIndex, err := manager.Get(baseURL+"%s/index.json", packageName)
+	if err != nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal,
+			fmt.Sprintf("failed to get nuget package registration index json: %v", err))
+	}
+	defer resPackageRegistrationIndex.Body.Close()
+	packageRegistrationCatalogRoot := &nugetPackageRegistrationCatalogRoot{}
+	err = json.NewDecoder(resPackageRegistrationIndex.Body).Decode(packageRegistrationCatalogRoot)
+	if err != nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal,
+			fmt.Sprintf("failed to parse package registration index json: %v", err))
+	}
+	return getLatestListedVersionFromPackageRegistrationPages(packageRegistrationCatalogRoot.Pages, manager)
```

Comment:
same comment about being able to define this as a method on `nugetPackageRegistrationCatalogRoot`

Maybe something like ?
```
func (n nugetPackageRegistrationCatalogRoot) LatestVersion( manager packageManagerClient)

Reply:
makes sense. fixed